---
title: "Solutions to Homework set 4"
output: pdf_document
---

```{r, include=FALSE}
# change output format
knitr::opts_chunk$set (prompt=TRUE, comment=NA)
```
Load library __genetics__, and datasets __FAMuSS__ and __HGDP__:
```{r message=FALSE}
library (genetics)
fmsURL <- "http://people.umass.edu/steinruecken/data/FMS_data.txt"
fms <- read.delim (file=fmsURL, header=T, sep="\t")
hgdpURL <- "http://people.umass.edu/steinruecken/data/HGDP_AKT1.txt"
hgdp <- read.delim (file=hgdpURL, header=T, sep="\t")
```


__Question 1__: Test whether there is a deviation from Hardy-Weinberg equilibrium (HWE) for the \texttt{akt1\_a7699g} SNP based on the full \texttt{FAMuSS} dataset, using Fisherâ€™s exact test. Check whether stratifying by the variable \texttt{Race} alters your findings.

__Answer__: 
```{r}
snpData <- genotype(fms$akt1_a7699g, sep="")
HWE.exact(snpData)$p.value
```

Testing the SNP in the complete sample significantly violates HWE ($\alpha=0.05$).

```{r}
tapply (snpData, INDEX=fms$Race,
        function (x) {HWE.exact(x)$p.value} )
```

However, within each Race, HWE is only violated in the Asian population ($\alpha=0.05$, No multiple testing correction). The Am. Indian population, however, has only one individual.

__Question 2__: Assess whether there is any evidence for genotyping errors in the __AKT1__ gene for the __HGDP__ data. To this end, perform tests for HWE ($\chi^2$-test) at each SNP for the whole sample and stratified by __Geographic.area__. What do the results indicate about possible genotyping errors?

__Answer__: Test for Hardy-Weinberg equilibrium (HWE) using $\chi^2$-test at each SNP:
```{r}
# get the names
set.seed (42)
aktNames <- names(hgdp)[(substr(names(hgdp),1,4)=="AKT1")]
pValues <- c()
# test at each SNP
for (i in 1:length(aktNames)) {
#   set.seed (42)
  # get SNP
  aktSnp <- genotype (hgdp[[aktNames[i]]], sep="")
  # get p-value using chi-square test for HWE
  pValue <- HWE.chisq(aktSnp)$p.value
  # store pValue
  pValues <- append (pValues, pValue)
}
aktNames
pValues
```
HWE is violated at each SNP, which would suggest genotyping erros at these SNPs.

However, we know that population substructure can influence HWE. Thus we perform the HWE tests at each SNP stratified by __Geographic.area__:
```{r}
set.seed (42)
# test each SNP
geoPValues <- c()
for (i in 1:length(aktNames)) {
#   set.seed (42)
  # get SNP
  aktSnp <- genotype (hgdp[[aktNames[i]]], sep="")
  # test for HWE stratified by geographic area
  hweGeoArea <- tapply (aktSnp, INDEX=hgdp$Geographic.area,
#                         function (x) {set.seed (42); HWE.chisq(x)$p.value} )
                        function (x) {HWE.chisq(x)$p.value} )
  print ("")
  print (paste("===============", aktNames[i], "==============="))
  print (hweGeoArea)
  # store pValue
  geoPValues <- append (geoPValues, hweGeoArea)
}
sort(geoPValues)[1:8]
```
Without multiple testing correction, we see evidence for violation of HWE in South America for all four SNPs, in Russia at SNP __AKT1.C0756A__ and __AKT1.C6024T__, and in Central Africa at SNP __AKT1.G2347T__. Thus we would conclude that there are potential genotyping errors at all four SNPs in South America, and at the indicated SNPs in Russia and Central Africa.

__Question 3__: Using the SNPs in the __akt1__ and __akt2__ genes in the __FAMuSS__ dataset: Determine whether there is any evidence for population substructure in African Americans. To this end, perform either a Multidimensional Scaling (MDS) or a Principal component analysis (PCA) analysis with two dimensions and explain what the result indicates about potential population structure.

__Answer__: First get the dataset of the SNPs in the two genes:
```{r} 
namesAktSnps <- names(fms)[(substr(names(fms),1,4)=="akt1") |
                             (substr(names(fms),1,4)=="akt2")]
# namesAktSnps <- names(fms)[(substr(names(fms),1,4)=="akt2")]
namesAktSnps
fmsAkt <- fms[,is.element(names(fms),namesAktSnps)]
```
Then extract the subset for __African Americans__:
```{r}
# be mindful of NA
afrAmAkt <- fmsAkt[!is.na(fms$Race) & fms$Race == "African Am",]
```
Create the numerical genotypes we need for either __MDS__ or __PCA__ (NA -> 4, but argueably 1,2,3 are possible as well)
```{r}
afrAmAktNum <- data.matrix (afrAmAkt)
afrAmAktNum[is.na(afrAmAkt)] <- 4
```

__Alternative 1: Multidimensional scaling (MDS)__

Compute the matrix of pairwise distances between all indiviuals and plot the first two coordinates from the MDS analysis:
```{r}
distAfrAmAkt <- as.matrix (dist (afrAmAktNum))
plot (cmdscale (distAfrAmAkt), xlab="C1", ylab="C2", main="MDS")
```

__Alternative 2: Principal component analysis (PCA)__

Do a principal component analysis and plot the projection of the data onto the first two principal components:
```{r}
PCAKT <- prcomp (afrAmAktNum)
plot (PCAKT[["x"]][,1], PCAKT[["x"]][,2],
      xlab="PC1", ylab="PC2", main="PCA")
```

__Conclusion__: Either in the MDS-plot or the PCA-plot, we see that there are two distinct clusters, which suggests that there is population substructure, even within the African Americans.

__Question 4__: Write an R script to determine the mean and variance of identical-by-state
(IBS) allele sharing between all pairs of individuals within a given sample. Apply it to the 24 SNPs in the __akt1__ gene in the __FAMuSS__ data. You only have to apply it to the first 20 individuals of the dataset($1 \leq i < j \leq 20$). Produce a scatterplot of the means and variances for the different pairs of individuals.

__Answer__: The following functions takes a data-frame of genotypes at a collection of SNPs for a sample of individuals and reutrns a list of pairs $(\text{mean IBS}, \text{ variance IBS}) = (\mu_{ij}, \sigma_{ij}^2)$ for all pairs of individuals $i < j$:
```{r}
# allele sharing at single SNP between two individuals
alleleSharing <- function (x, naValue=NA) {
  # NA's give no in common
  if (is.na(x[1]) || is.na(x[2])) {
    return (naValue)
  }
  else {
    if (x[1] == x[2]) {
      # 2 in common
      return (2)
    }
    else {
      # count how large the intersect
      chars <- strsplit(as.character(x),split='')
      return (length (intersect (chars[[1]],chars[[2]])))
    }
  }
}

# IBS function for a given dataset
IBS <- function (genoFrame, naValue=NA) {
  # how many individuals
  nInd <- dim(genoFrame)[1]
  # for through all pairs
  results <- c()
  for (i in 1:(nInd-1)) {
    for (j in (i+1):nInd) {
      # join the two in question together
      joined <- rbind (genoFrame[i,], genoFrame[j,])
      # determine number of alleles in common at each SNP
      ibs <- apply (joined, 2, function (x) { alleleSharing (x, naValue) } )
      # remeber mean and variance
      results <- append (results, c(mean(ibs,na.rm=TRUE), var(ibs,na.rm=TRUE)))
    }
  }
  # return it (transposed)
  return (t (matrix (results,nrow=2)))
}

```
Now create a data-frame that has the genotypes for all individuals from the __FAMuSS__ data at the SNPs in the __akt1__ gene:
```{r}
namesAkt1Snps <- names(fms)[substr(names(fms),1,4)=="akt1"]
fmsAkt1 <- fms[,is.element(names(fms),namesAkt1Snps)]
```
Apply our function to get the means and variances of IBS in between all pairs of individuals:
```{r}
ibsStat <- IBS (fmsAkt1[1:20,], naValue=0)
```
Here we use 0 as the IBS value at SNPs were at least one genotype is missing.

And make a scatterplot with these values:
```{r}
plot (ibsStat[,1], ibsStat[,2], xlab='mean IBS', ylab='variance IBS', main="Scatterplot IBS")
```

We see some pathologies, which is due to the small number of SNPs we used. However, there is a large group with low mean and high variance (unrelated), but also some pairs of individuals with high mean and low variance (related).

__Alternative__: Use __NA__ instead of __0__ whenever genotypes are missing:
```{r}
ibsStat <- IBS (fmsAkt1[1:20,], naValue=NA)
plot (ibsStat[,1], ibsStat[,2], xlab='mean IBS', ylab='variance IBS', main="Scatterplot IBS")
```

Again, most pairs have low mean and high variance. Howeevr, a few pairs have high mean IBS and low variance, which is evidence for relatedness.

__Question 5__: Define and contrast the following terms: (1) family-wise error, (2) type-1 and
type-2 errors, (3) complete and partial nulls, and (4) strong and weak controls.

__Answer__: In what follows, consider testing the $m$ null hypotheses $H_0^1,\ldots,H_0^m$:

1. The family-wise error rate (FWER) is the probability of making at least one type-1 error (see 2.) among the $m$ tests:
$$ FWER = P(V\geq 1) $$
where $V$ is the number of type-1 errors.

2. __Type-1 error__: Rejecting a null hypothesis $H_0^i$, although it is true, and should thus not be rejected. __Type-2 error__: Accepting  a null hypothesis $H_0^i$, although it is false (the alternative hypothesis is true), and it should not be accepted.

3.\& 4. The FWER can be controlled in a weak or a strong sense. To this end, define the FWE under the complete null (FWEC) as
$$
  FWEC = P(V\geq 1 | \text{ all $H_0^i$ true}) = P(V\geq 1 | \text{ $H_0^C$ true}),
$$
with $H_0^C = \{H_0^1,\ldots,H_0^m\}$ being the collection of all null hypotheses (the __complete null__). Controlling the FWER in the __weak__ sense means that we can guarantee that $\text{FWEC} \leq \alpha$ for a given level $\alpha$. 
Furthermore, define the FWE under a partial null (FWEP) as
$$
  FWEP = P(V\geq 1 | \text{ given subset $\{H_0^{i_1},\ldots,H_0^{i_k}\}$ true})
$$
where $\{H_0^{i_1},\ldots,H_0^{i_k}\} \subset \{H_0^1,\ldots,H_0^m\}$ is a given subset of null hypotheses (__partial null__). Controlling the FWER in the __strong__ sense means that we can guarantee that $\text{FWEP} \leq \alpha$ for __all__ possible sets of partial nulls (including the empty set). Usually, controlling in the strong sense is desirable, however, more difficult to achieve.

