---
title: "690T Final Project"
author: "Aixa Perez-Caraballo, Meilan Chen, Alexandra Purdue-Smithe"
date: "April 14, 2016"
output: pdf_document
---


3 traits: Asthma, Irritable_bowel_syndrome, Colon_polyps

## Question 1

```{r}
library(genetics)
library (LDheatmap)
# load data
data <- read.table("/Users/meilanchen/Google Drive/690T/Group1.data",header=T, sep="\t")
# remove columns doesn't have full snps and create the new dataset
rs.names <- names (data)[substr(names(data),1,2)== "rs"]
rs.snps <- makeGenotypes(data[rs.names], sep ="")
newsnps <- rs.snps[, sapply(rs.snps, function(col)length(unique(col))) > 3]
other.names <- names (data)[substr(names(data),1,2)!= "rs"]
other <- data[other.names]
newdata <- data.frame(newsnps, other)



```



## Question 2 

**Principal component analysis (PCA)**. Use PCA on the genotype data to determine whether there is evidence for population sub-structure in the data. Explain how you reached this conclusion. Assign one of four sub-population labels to each individual as follows: Choose a threshold each for principal component 1 and principal component 2 such that each individual is assigned to one of the four regions in the two dimensional space defined by these thresholds. Use the regions as labels. Choose the thresholds such that the regions and the associated labels reflect the sub-structure in the dataset. You might find that 3 labels are sufficient.

```{r}

rs.genoNum <- data.matrix(newsnps)
rs.genoNum[is.na(rs.genoNum)] <- 4
PCA.rs <- prcomp(rs.genoNum)

plot(PCA.rs$"x"[,1],PCA.rs$"x"[,2],xlab="PC1",ylab="PC2")
abline(v=50,lty=2)
abline(h=30,lty=2)

partition <- ifelse(PCA.rs$"x"[, 1] < 50 & PCA.rs$"x"[, 2] < 30, "region one", 
             ifelse(PCA.rs$"x"[, 1] > 50 & PCA.rs$"x"[, 2] > 30, "region two",
             ifelse(PCA.rs$"x"[, 1] > 50 & PCA.rs$"x"[, 2] < 30, "region three", "other")))


```


**Answer:** With three distinct regions with clear clusters of people, we conclude that there is evidence of population substructure when using PCA. 

## Question 3 

**Hardy-Weinberg Equilibrium (HWE)**. Test for violation of HWE at each of the first 1000 SNPs of the dataset. Stratify the analysis by the population labels that you assigned in Question 2. Correct for multiple testing using the Benjamini-Hochberg procedure. Is there evidence for violation of HWE at any of the SNPs in any population?

```{r}

hwe.data <- as.data.frame(cbind(newsnps, partition))
hwe.region1.df <- hwe.data[which(hwe.data$partition == "region one"), ]
hwe.region2.df <- hwe.data[which(hwe.data$partition == "region two"), ]
hwe.region3.df <- hwe.data[which(hwe.data$partition == "region three"), ]

hwe.region1<-NA
for(i in 1:1000){
  set.seed(1)
  column1 <- hwe.region1.df[, i]
  p.value.hwe <- HWE.exact(column1)$p.value
  adj.pvalue.hwe <- p.adjust(p.value.hwe, method = "BH")
  hwe.region1[i] <- adj.pvalue.hwe
}


hwe.region2<-NA
for(i in 1:1000){
  set.seed(1)
  column2 <- hwe.region2.df[, i]
  p.value.hwe<-HWE.exact(column2)$p.value
  adj.pvalue.hwe <- p.adjust(p.value.hwe, method = "BH")
  hwe.region2[i] <- adj.pvalue.hwe
}

hwe.region3<-NA
for(i in 1:1000){
  set.seed(1)
  column3 <- hwe.region3.df[, i]
  p.value.hwe<-HWE.exact(column3)$p.value
  adj.pvalue.hwe <- p.adjust(p.value.hwe, method = "BH")
  hwe.region3[i] <- adj.pvalue.hwe
}


sum(hwe.region1<0.05)
sum(hwe.region2<0.05)
sum(hwe.region3<0.05)

```

**Answer:** There are 57 snps association in region one, where there is a level of significance of less than 0.05. And there are 6 snps association in region two. In region three, we saw zero significant snps for a test of less than 0.05. 

## Question 4 

**Linkage disequilibrium (LD)**. Use the first 50 SNPs of the dataset and compute a matrix of pairwise LD (measure D'). Furthermore, plot the pairwise LD structure using the function LDheatmap, and interpret the results.


```{r, fig.height=8}

# sunbstract data
ld.data <- newsnps[1:50]

# Matrix of pairwise D' values
LD <- round(LD (ld.data)[["D'"]], 2)

# Visualization of LD
LDheatmap(ld.data, LDmeasure="D'")

```


## Question 5 

**Association: Genotype - Binary Trait.** For each of the 3 binary disease traits of your choice, test whether any of the SNPs in the dataset is significantly associated with the trait. Correct for multiple testing using the Bonferroni adjustment, and separately, also the Benjamini-Hochberg adjustment. Perform the multiple testing adjustment for each binary trait separately. Compare the different methods correcting for multiple testing. Interpret your results.

```{r}

####################################################################
################   fisher.test     ##################################
####################################################################

trait1 <- newdata$Irritable_bowel_syndrome
trait2 <- newdata$Asthma
trait3 <- newdata$Colon_polyps

fischert1 <- function(Geno){
  ObsTab <- table(trait1, Geno)
  return(fisher.test(ObsTab)$p.value)
}
fischert2 <- function(Geno){
  ObsTab <- table(trait2, Geno)
  return(fisher.test(ObsTab)$p.value)
}
fischert3 <- function(Geno){
  ObsTab <- table(trait3, Geno)
  return(fisher.test(ObsTab)$p.value)
}

p.values1 <- apply(newsnps, 2, fischert1)
p.values2 <- apply(newsnps, 2, fischert2)
p.values3 <- apply(newsnps, 2, fischert3)


p1 <- names(newsnps)[p.values1< 0.05]
p2 <- names(newsnps)[p.values2< 0.05]
p3 <- names(newsnps)[p.values3< 0.05]

length(p1)
length(p2)
length(p3)

adj.pvalue1.bon <- p.adjust(p.values1, method = "bonferroni")
adj.pvalue2.bon <- p.adjust(p.values2, method = "bonferroni")
adj.pvalue3.bon <- p.adjust(p.values3, method = "bonferroni")

p1.bon <- names(newsnps)[adj.pvalue1.bon< 0.05]
p2.bon <- names(newsnps)[adj.pvalue2.bon< 0.05]
p3.bon <- names(newsnps)[adj.pvalue3.bon< 0.05]

length(p1.bon)
length(p2.bon)
length(p3.bon)

adj.pvalue1.bh <- p.adjust(p.values1, method = "BH")
adj.pvalue2.bh  <- p.adjust(p.values2, method = "BH")
adj.pvalue3.bh <- p.adjust(p.values3, method = "BH")

p1.bh <- names(newsnps)[adj.pvalue1.bh< 0.05]
p2.bh <- names(newsnps)[adj.pvalue2.bh< 0.05]
p3.bh <- names(newsnps)[adj.pvalue3.bh< 0.05]

length(p1.bh)
length(p2.bh)
length(p3.bh)

####################################################################
################   chisq.test     ##################################
####################################################################

# 
# trait1 <- newdata$Irritable_bowel_syndrome
# trait2 <- newdata$Asthma
# trait3 <- newdata$Colon_polyps
# 
# chisq.test1 <- function(Geno){
#   ObsTab <- table(trait1, Geno)
#   return(chisq.test(ObsTab)$p.value)
# }
# chisq.test2 <- function(Geno){
#   ObsTab <- table(trait2, Geno)
#   return(chisq.test(ObsTab)$p.value)
# }
# chisq.test3 <- function(Geno){
#   ObsTab <- table(trait3, Geno)
#   return(chisq.test(ObsTab)$p.value)
# }
# 
# p.values1 <- apply(newsnps, 2, chisq.test1)
# p.values2 <- apply(newsnps, 2, chisq.test2)
# p.values3 <- apply(newsnps, 2, chisq.test3)
# 
# names(newsnps)[p.values1< 0.05]
# names(newsnps)[p.values2< 0.05]
# names(newsnps)[p.values3< 0.05]
# 
# adj.pvalue1.bon <- p.adjust(p.values1, method = "bonferroni")
# adj.pvalue2.bon <- p.adjust(p.values2, method = "bonferroni")
# adj.pvalue3.bon <- p.adjust(p.values3, method = "bonferroni")
# 
# names(newsnps)[adj.pvalue1.bon< 0.05]
# names(newsnps)[adj.pvalue2.bon< 0.05]
# names(newsnps)[adj.pvalue3.bon< 0.05]
# 
# adj.pvalue1.bh <- p.adjust(p.values1, method = "BH")
# adj.pvalue2.bh  <- p.adjust(p.values2, method = "BH")
# adj.pvalue3.bh <- p.adjust(p.values3, method = "BH")
# 
# names(newsnps)[adj.pvalue1.bh< 0.05]
# names(newsnps)[adj.pvalue2.bh< 0.05]
# names(newsnps)[adj.pvalue3.bh< 0.05]
# adj.qvalue1 <- qvalue(p.values1, pi0.method = "bootstrap")$qvalues
# adj.qvalue2 <- qvalue(p.values2, pi0.method = "bootstrap")$qvalues
# adj.qvalue3 <- qvalue(p.values3, pi0.method = "bootstrap")$qvalues
# 
# names(newsnps)[adj.qvalue1 < 0.05]
# names(newsnps)[adj.qvalue1 < 0.05]
# names(newsnps)[adj.qvalue1 < 0.05]


```

**Answer:** After correction there is no significant association. 

## Question 6

**Effective number of tests.** Using the first 100 SNPs in the dataset, determine
the effective number of tests ($M_{eff}$ and $M_{eff-G}$ with C = 0.95) if we were to test for association with a trait. Encode the genotypes numerically as the number of minor alleles (set NA to 1). Use the absolute value of Pearsons correlation coefficient as a measure of correlation between two SNPs.

```{r}

snps.6 <- newsnps[,1:100]

# turn into number minor alleles
numMinor <- function (x) {
  # get a numerical representation
  # this counts number of copies of lexicographically larger allele
  numGeno <- as.numeric (as.factor(x)) - 1
  # if lexicographically larger allele is minor allele, then change nothing, otherwise reverse
  alleles <- allele.names(genotype (as.character(x), sep="", reorder="freq"))
  if (alleles[1] > alleles[2]) {
    numGeno <- 2 - numGeno
  }
  # get rid of NA
  numGeno[is.na(numGeno)] <- 1
  # return it
  return (numGeno)
}
# transform genotype data into numeric data
names.Snps <- names(snps.6)
snps.6rs <- snps.6[,is.element(names(snps.6),names.Snps)]
rs.Num <- apply (snps.6rs, 2, numMinor)
# compute eigenvalues
V <- abs (cor (rs.Num))
eValues <- eigen(V)$values
# normal M_eff
m <- length(eValues)
M.eff <- 1 + (m-1) * (1 - var(eValues)/m)
M.eff
# and M_eff-C
thC <- 0.95
M.effG <- min(which(cumsum(eValues)/sum(eValues) > thC))
M.effG 


```


## Question 7

**Odds ratio**. For each of the 3 binary traits, determine the 10 SNPs with the lowest p-value from Question 5. For each of these SNPs, determine the odds ratio of having the disease trait for individuals that are homozygous for the major allele compared to the individuals having at least one minor allele.

```{r}
trait1 <- newdata$Irritable_bowel_syndrome
trait2 <- newdata$Asthma
trait3 <- newdata$Colon_polyps
trait1.p <- sort(p.values1)
trait2.p <- sort(p.values2)
trait3.p <- sort(p.values3)
Snps.7.1 <- names(trait1.p[1:10])
Snps.7.2 <- names(trait2.p[1:10])
Snps.7.3 <- names(trait3.p[1:10])


Geno.1 <- makeGenotypes(data[Snps.7.1])
Geno.2 <- makeGenotypes(data[Snps.7.2])
Geno.3 <- makeGenotypes(data[Snps.7.3])

# first trait

t1 <- list()
maj.1 <- rep(NA, 10)
OR.trait1 <- rep(NA,10)

for (i in 1:10){
        t1[[i]] <- table(trait1, Geno.1[,i])
        maj.1[i] <- which.max(c(apply(t1[[i]], 2, sum)[1], apply(t1[[i]], 2, sum)[3]))
       OR.trait1[i] <- if (maj.1[i] == 1 ) (t1[[i]][1,1]*(t1[[i]][2,2]+t1[[i]][2,3]))/(t1[[i]][2,1]*(t1[[i]][1,2]+t1[[i]][1,3]))
        else (t1[[i]][1,3]*(t1[[i]][2,2]+t1[[i]][2,1]))/(t1[[i]][2,3]*(t1[[i]][1,2]+t1[[i]][1,1]))
}

OR.trait1

# second trait

t2 <- list()
maj.2 <- rep(NA, 10)
OR.trait2 <- rep(NA,10)

for (i in 1:10){
        t2[[i]] <- table(trait2, Geno.2[,i])
        maj.2[i] <- which.max(c(apply(t2[[i]], 2, sum)[1], apply(t2[[i]], 2, sum)[3]))
       OR.trait2[i] <- if (maj.2[i] == 1 ) (t2[[i]][1,1]*(t2[[i]][2,2]+t2[[i]][2,3]))/(t2[[i]][2,1]*(t2[[i]][1,2]+t2[[i]][1,3]))
        else (t2[[i]][1,3]*(t2[[i]][2,2]+t2[[i]][2,1]))/(t2[[i]][2,3]*(t2[[i]][1,2]+t2[[i]][1,1]))
}

OR.trait2

# Third trait

t3 <- list()
maj.3 <- rep(NA, 10)
OR.trait3 <- rep(NA,10)

for (i in 1:10){
        t3[[i]] <- table(trait3, Geno.3[,i])
        maj.3[i] <- which.max(c(apply(t3[[i]], 2, sum)[1], apply(t3[[i]], 2, sum)[3]))
       OR.trait3[i] <- if (maj.3[i] == 1 ) (t3[[i]][1,1]*(t3[[i]][2,2]+t3[[i]][2,3]))/(t3[[i]][2,1]*(t3[[i]][1,2]+t3[[i]][1,3]))
        else (t3[[i]][1,3]*(t3[[i]][2,2]+t3[[i]][2,1]))/(t3[[i]][2,3]*(t3[[i]][1,2]+t3[[i]][1,1]))
}

OR.trait3


```


## Question 8 

**Association: Genotype - Quantitative Trait**. Determine whether any of the first 500 SNPs in the dataset is associated with either of the quantitative traits Weight or Height (investigate both traits). At each SNP, use an indicator variable for the presence of at least one minor allele, and perform a t-test for association. Correct for multiple testing using the Bonferroni adjustment and (separately) the Benjamini-Hochberg adjustment. Apply the correction to each quantitative trait separately. Interpret your results.

```{r}
rs.names <- names (data)[substr(names(data),1,2)== "rs"]
rs.snps <- makeGenotypes(data[rs.names])
newsnps <- rs.snps[, sapply(rs.snps, function(col)length(unique(col))) > 3]
Snps.8 <- newsnps[, 1:500]

#Associations with height
tTestPval <- function(Geno) {
        majorAllele <- allele.names (genotype (Geno, sep="", reorder="freq"))[1]
        GenoMaj <- paste (majorAllele, majorAllele, sep="")
        GenoBin <- as.numeric(Geno==GenoMaj)[!is.na(Geno)]
        Trait <- newdata$Height[!is.na(Geno)]
        return (t.test (Trait[GenoBin==1],Trait[GenoBin==0])$p.value)
}

p.values1 <- apply(Snps.8, 2, tTestPval)
p1 <- names(newsnps)[p.values1< 0.05]
length(p1)
adj.pvalue1.bon <- p.adjust(p.values1, method = "bonferroni")
p1.bon <- names(newsnps)[adj.pvalue1.bon< 0.05]
length(p1.bon)

adj.pvalue3.bh <- p.adjust(p.values1, method = "BH")
p1.bh <- names(newsnps)[adj.pvalue1.bh< 0.05]
length(p1.bh)



#Associations with height
tTestPval.2 <- function(Geno) {
        majorAllele <- allele.names (genotype (Geno, sep="", reorder="freq"))[1]
        GenoMaj <- paste (majorAllele, majorAllele, sep="")
        GenoBin <- as.numeric(Geno==GenoMaj)[!is.na(Geno)]
        Trait <- newdata$Weight[!is.na(Geno)]
        return (t.test (Trait[GenoBin==1],Trait[GenoBin==0])$p.value)
}

p.values2 <- apply(Snps.8, 2, tTestPval.2)
p2 <- names(newsnps)[p.values2< 0.05]
length(p2)
adj.pvalue2.bon <- p.adjust(p.values2, method = "bonferroni")
p2.bon <- names(newsnps)[adj.pvalue2.bon< 0.05]
length(p2.bon)

adj.pvalue2.bh <- p.adjust(p.values2, method = "BH")
p2.bh <- names(newsnps)[adj.pvalue2.bh< 0.05]
length(p2.bh)



```

**Answer:** After correction there is no significant association.